### PR TITLE
In-code key handling system, Take 2

### PR DESCRIPTION
**Pull request by ninjanomnom**
:cl: MrPerson & ninjanomnom
refactor: Completely changed how keyboard input is read.
add: Holding two directions at the same time will now move you diagonally. This works with the arrow keys, wasd, and the numpad.
tweak: Moving diagonally takes twice as long as moving a cardinal direction.
fix: You can use control to turn using wasd and the numpad instead of just the arrow keys.
del: Some old non-hotkey mode behaviors, especially in relation to chatbox interaction, can't be kept with the new system. Of key note: You can't type while walking. This is due to limitations of byond and the work necessary to overcome it is better done as another overhaul allowing custom controls.
/:cl:

An updated version of #18863 by @MrPerson 

Ready for testmerging and I'll be working out any other issues we have with it.

Video of it working https://puu.sh/ylF1Q/1285840bab.mp4

### CODE REVIEW

**Comment by PKPenguin321** [comment]
Oh nice, I had forgotten about this. Hope it will go through this time!

**Comment by WJohn** [comment]
Does diagonal moving into a wall still have slowed movement or do you regain full speed considering you're not moving at an angle?

Also it would be nice if simple animals had slowed diagonal movement to better mimic a player.

**Comment by ninjanomnom** [comment]
You don't have slowed movement while sliding across a wall. Simple animals are slowed down from this already I think but I'll check.

**Comment by Qustinnus** [comment]
This is awesome, I'd love to see control settings being a configurable thing in the future, too.

**Comment by RandomMarine** [comment]
2017-11-14 20_11_42-space station 13
Is it possible to slip through gaps like these with diagonal movement?

**Comment by Cyberboss** [comment]
@MrPerson had a more recent PR, did you see that one?

File: code/modules/keybindings/bindings_carbon.dm [hunk 1]
```diff
@@ -0,0 +1,19 @@
+/mob/living/carbon/key_down(_key, client/user)
+	switch(_key)
+		if("numpad1", "southwest", "r")
+			toggle_throw_mode()
+		if("numpad7", "northwest", "q")
+			var/obj/item/I = get_active_held_item()
+			if(!I)
+				src << "<span class='warning'>You have nothing to drop in your hand!</span>"
```
Comment (Cyberboss) [inline code/modules/keybindings/bindings_carbon.dm]: Reee

File: code/modules/keybindings/bindings_atom.dm [hunk 2]
```diff
@@ -0,0 +1,18 @@
+// You might be wondering why this isn't client level. If focus is null, we don't want you to move.
+// Only way to do that is to tie the behavior into the focus's keyLoop().
+
+/atom/movable/keyLoop(client/user)
+	if(!user.keys_held["ctrl"])
+		var/movement_dir = 0
```
Comment (Cyberboss) [inline code/modules/keybindings/bindings_atom.dm]: NONE

File: code/modules/keybindings/focus.dm [hunk 3]
```diff
@@ -0,0 +1,20 @@
+/datum
+	var/list/focusers //Only initialized when needed. Contains a list of mobs focusing on this.
```
Comment (Cyberboss) [inline code/modules/keybindings/focus.dm]: datum.dm

File: code/modules/keybindings/focus.dm [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+/mob
+	var/datum/focus //What receives our keyboard inputs. src by default
```
Comment (Cyberboss) [inline code/modules/keybindings/focus.dm]: mob_defines.dm

File: code/modules/keybindings/setup.dm [hunk 5]
```diff
@@ -0,0 +1,27 @@
+/client
+	var/list/keys_held = list() // A list of any keys held currently
+	// These next two vars are to apply movement for keypresses and releases made while move delayed.
+	// Because discarding that input makes the game less responsive.
+	var/next_move_dir_add // On next move, add this dir to the move that would otherwise be done
+	var/next_move_dir_sub // On next move, subtract this dir from the move that would otherwise be done
```
Comment (Cyberboss) [inline code/modules/keybindings/setup.dm]: client_defines.dm

File: code/modules/mob/living/carbon/alien/humanoid/caste/hunter.dm [hunk 6]
```diff
@@ -11,7 +11,7 @@
 	..()
 
 /mob/living/carbon/alien/humanoid/hunter/movement_delay()
-	. = -1		//hunters are sanic
+	. = -0.3	//hunters are sanic
```
Comment (Cyberboss) [inline code/modules/mob/living/carbon/alien/humanoid/caste/hunter.dm]: REEEEEEE

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Comment by ninjanomnom** [comment]
@RandomMarine no

File: code/modules/mob/living/carbon/alien/humanoid/caste/hunter.dm (hunk 6 above)
Comment (ninjanomnom) [reply code/modules/mob/living/carbon/alien/humanoid/caste/hunter.dm]: some things got moved over incorrectly from mrperson's pr, or this was in the original for some reason

File: code/modules/mob/say.dm [hunk 7]
```diff
@@ -5,7 +5,10 @@
 	if(GLOB.say_disabled)	//This is here to try to identify lag problems
 		to_chat(usr, "<span class='danger'>Speech is currently admin-disabled.</span>")
 		return
-	usr.say(message)
+	if(!message)
```
Comment (AnturK) [inline code/modules/mob/say.dm]: Personal preference but this is going to be  bit annoying.

**Comment by AnturK** [comment]
>Ready to testmerge
>Hotkey mode removed

**Comment by ninjanomnom** [comment]
hotkey mode works just fine from what I can tell? what's happening when you press tab?

**Comment by MrStonedOne** [comment]
Hotkey mode issues are a byond bug

**Comment by epochayur** [comment]
>add: Holding two directions at the same time will now move you diagonally

never thought I'd see the day

File: code/modules/keybindings/setup.dm [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+Aparently you need to wrap the entire proc + args in quotes if you intend on using args
+But you don't need the quote wrappings to just call on a proc with no args
+ex. winset(src, "default-Any", "command=keyDown \[\[*\]\]") 		fail: command = keyDown
+ex. winset(src, "default-Any", "command=keyDown \"\[\[*\]\]\"") 	fail: same
+ex. winset(src, "default-T", "command=say") 						works fine
+ex. winset(src, "default-Any", "command=\"keyDown \[\[*\]\]\"")		works fine
+Thanks for the useful errors lummox ~ninjanomnom
+*/
+GLOBAL_LIST_INIT(default_macros, list(
+	"Tab" = "\".winset \\\"input.focus=true?map.focus=true input.background-color=#F0F0F0:input.focus=true input.background-color=#D3B5B5\\\"\"",
```
Comment (ninjanomnom) [inline code/modules/keybindings/setup.dm]: This is my magnum opus of shitcode

**Comment by CosmicScientist** [comment]
Server revision compiled on: 2017-11-20
The following pull requests are currently test merged:
#32751: 'In-code key handling system, Take 2' by ninjanomnom at commit ce5271
Based off origin/master commit: e79626
Server tools version: /tg/station 13 Server v3.2.0.12
Server tools API version: 3.2.0.1

reeeee
Hotkeys are broken. R has no functionality (E has its functionality) so you have to resort to using End to hotkey throw objects _so you don't equip/stow what you want to throw._ I don't know what other hotkeys have functionality, combined or removed.

Pressing tab to swap between hotkey and arrowkey mode is broken, in that if you're in arrowkey mode (verb bar highlighted) you can still use all keys without input defaulting to the verb bar (so, WASD walk about and hotkeys) ___if___ you click first somewhere else after swapping to arrowkey mode (who isn't going to be clicking in this game?)

Having Chat by default enabled makes it impossible to use backspace to remove the _say"_ letters in case you wanted to write a verb (e.g. suicide, ghost, rest, sleep) instead of speech or emote.

File: code/modules/keybindings/bindings_carbon.dm [hunk 9]
```diff
@@ -0,0 +1,19 @@
+/mob/living/carbon/key_down(_key, client/user)
+	switch(_key)
+		if("Numpad1", "Southwest", "E")
+			toggle_throw_mode()
```
Comment (CosmicScientist) [inline code/modules/keybindings/bindings_carbon.dm]: Toggle throw needs to be "R"

**Comment by CosmicScientist** [comment]
Numpad body target selection and borg module hotkeys are also apparently frakked up according to players.

File: code/modules/client/client_procs.dm [hunk 10]
```diff
@@ -203,7 +203,6 @@ GLOBAL_LIST(external_rsc_urls)
 	prefs.last_id = computer_id			//these are gonna be used for banning
 	if(world.byond_version >= 511 && byond_version >= 511 && prefs.clientfps)
 		vars["fps"] = prefs.clientfps
-	sethotkeys(1)						//set hoykeys from preferences (from_pref = 1)
```
Comment (MrStonedOne) [inline code/modules/client/client_procs.dm]: This is the source of all tab hotkey mode bugs.

File: code/modules/keybindings/bindings_carbon.dm (hunk 9 above)
Comment (CosmicScientist) [inline code/modules/keybindings/bindings_carbon.dm]: Numpad 1-8 are for body targeting.

File: code/modules/keybindings/bindings_carbon.dm [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+		if("Numpad7", "Northwest", "Q")
```
Comment (CosmicScientist) [inline code/modules/keybindings/bindings_carbon.dm]: Numpad 1-8 are for body targeting.

File: code/modules/keybindings/bindings_mob.dm [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+		if("Delete")
+			if(!pulling)
+				to_chat(src, "<span class='notice'>You are not pulling anything.</span>")
+			else
+				stop_pulling()
+		if("Insert", "G")
+			a_intent_change("Right")
+		if("F")
+			a_intent_change("Left")
+		if("Numpad9", "northeast", "X")
```
Comment (CosmicScientist) [inline code/modules/keybindings/bindings_mob.dm]: Numpad 1-8 are for body targeting.

File: code/modules/keybindings/bindings_mob.dm [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+			swap_hand()
+		if("Numpad3", "Southeast", "Y", "Z") // attack_self(). No idea who came up with "mode()"
```
Comment (CosmicScientist) [inline code/modules/keybindings/bindings_mob.dm]: Numpad 1-8 are for body targeting.

File: code/modules/keybindings/setup.dm [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+	return
+/datum/proc/keyLoop(client/user) // Called once every frame
+	return
+
+// Keys used for movement
+GLOBAL_LIST_INIT(movement_keys, list(
+	"W" = NORTH, "A" = WEST, "S" = SOUTH, "D" = EAST,									// WASD
+	"North" = NORTH, "West" = WEST, "South" = SOUTH, "East" = EAST,						// Arrow keys
+	"Numpad8" = NORTH, "Numpad4" = WEST, "Numpad2" = SOUTH, "Numpad6" = EAST			// Numpad
+	))
```
Comment (CosmicScientist) [inline code/modules/keybindings/setup.dm]: I don't know if this affects the _Numpad 1-8 are for body targeting._

**Review by ghost — CHANGES_REQUESTED** [review]
Numpad 1-8 are for body targeting.

File: code/modules/keybindings/setup.dm [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+	if(prefs.hotkeys)
+		winset(src, null, "mapwindow.map.focus=true input.background-color=#e0e0e0")
+	else
+		winset(src, null, "input.focus=true input.background-color=#d3b5b5")
+
+// This was in the library I based this off of and I'm leaving it here in case someone cares
+// var/list/numpad_mappings = list("numpad0" = "0", "numpad1" = "1", "numpad2" = "2", "numpad3" = "3", "numpad4" = "4", "numpad5" = "5", "numpad6" = "6", "numpad7" = "7", "numpad8" = "8", "numpad9" = "9", "divide" = "/", "multiply" = "*", "subtract" = "-", "add" = "+", "decimal" = ".")
+// It may be useful to turn numpad input to regular input. If so uncomment the above line and just do
+//	if(numpad_mappings[key])
+//		key = numpad_mappings[key]
```
Comment (CosmicScientist) [inline code/modules/keybindings/setup.dm]: What use is this here? Is it to do with the above code? Would it be better in the readme?

**Comment by CosmicScientist** [comment]
In-game hotkey list:
Hotkey-Mode: (hotkey-mode must be on)
        TAB = toggle hotkey-mode
        a = left
        s = down
        d = right
        w = up
        q = drop
        e = equip
        r = throw
        m = me
        t = say
        o = OOC
        b = resist
        h = stop pulling
        x = swap-hand
        z = activate held object (or y)
        f = cycle-intents-left
        g = cycle-intents-right
        1 = help-intent
        2 = disarm-intent
        3 = grab-intent
        4 = harm-intent
        Numpad = Body target selection (Press 8 repeatedly for Head->Eyes->Mouth)
        Alt(HOLD) = Alter movement intent



Any-Mode: (hotkey doesn't need to be on)
        Ctrl+a = left
        Ctrl+s = down
        Ctrl+d = right
        Ctrl+w = up
        Ctrl+q = drop
        Ctrl+e = equip
        Ctrl+r = throw
        Ctrl+b = resist
        Ctrl+h = stop pulling
        Ctrl+o = OOC
        Ctrl+x = swap-hand
        Ctrl+z = activate held object (or Ctrl+y)
        Ctrl+f = cycle-intents-left
        Ctrl+g = cycle-intents-right
        Ctrl+1 = help-intent
        Ctrl+2 = disarm-intent
        Ctrl+3 = grab-intent
        Ctrl+4 = harm-intent
        Ctrl+'+/-' OR
        Shift+Mousewheel = Ghost zoom in/out
        DEL = stop pulling
        INS = cycle-intents-right
        HOME = drop
        PGUP = swap-hand
        PGDN = activate held object
        END = throw
        Ctrl+Numpad = Body target selection (Press 8 repeatedly for Head->Eyes->Mouth)

I do not understand how throw mode can be activated by End when I can't see it in your code but page up and I assume the other four verbose buttons nearby probably don't work either.

Could you resolve these hotkeys before doing another test merge, pretty please? The hotkey list can be found on the OOC tab or by typing in hotkeys-help

**Comment by epochayur** [comment]
I can't speak for other people, but I use the HOME and END shortcuts pretty extensively, I'd really hate to see those go

**Comment by nicbn** [comment]
Hotkeys not working:

* r for throw (e is triggering it instead)
* m for me
* f and g for intent selection
* numpad for body selection
* Holding alt for walking
* h to stop pulling
* b to resist

**Comment by Hatfish** [comment]
Chat mode is broken, you can't clear "say" with backspace and need to shut off chat mode manually.  Any click on the screen deselects the chat bar, forcing you to reselect it manually (otherwise pressing certain letters brings up their hotkey variants, like t for chat).  The page up key is broken as well, which is awful if you need to quickly switch hands.

**Comment by ninjanomnom** [comment]
I haven't really messed with the hotkeys themselves yet because I was ensuring it was functional, now that the workaround works for the byond bug I can start actual work on the hotkeys.

File: code/modules/keybindings/setup.dm (hunk 15 above)
Comment (ninjanomnom) [reply code/modules/keybindings/setup.dm]: leftover stuff from MrPerson's pr, I haven't touched 90% of his work yet

**Comment by imsxz** [comment]
Diagonal movement while ventcrawling pops you out of the pipes, no matter where you are (and also crashed me once). Sorry if this has already been mentioned

**Comment by ninjanomnom** [comment]
That's actually a bug we've had for a while with diagonal moves but you couldn't do it yourself before.

**Comment by ninjanomnom** [comment]
I've tried a couple ways in local testing to deal with not being able to type input while doing other actions like walking etc but unfortunately it doesn't seem possible to deal with it in a clean way without basically being forced to support both systems at once.

I'm leaving that functionality behind in exchange for the code improvements and new features like diagonal movement this allows as well as the held keys system which is the main reason I started on all this in the first place.

@KorPhaeron This is ready assuming you're fine with the above

**Pull request merged**